File: AoC/MMXV/haskell/day10.hs
include_bytes :: IO String
include_bytes = readFile "../puzzles/day10.txt"

look_and_say :: String -> String
look_and_say = lookup_until ""

lookup_until [] ( n : r ) = lookup_until [n] r
lookup_until buffer @ ( last : _ ) ( n : r ) | last == n = lookup_until ( n : buffer ) r
lookup_until buffer @ ( last : _ ) ( n : r ) | last /= n = (say buffer) ++ lookup_until [n] r
lookup_until buffer [] = say buffer

say :: String -> String
say p = (show $ length p) ++ [(head p)]

-- n times
r' :: Int -> (a -> a) -> a -> a
r' 0 _ p = p
r' n f p = f $ r' (n - 1) f p
t' :: Int -> String -> String
t' n input = r' n look_and_say input

-- Main

main :: IO ()
main = do
    input <- include_bytes

    let s = input

    putStrLn $ "--- Day 10: Elves Look, Elves Say ---"

    putStrLn $ "--- Part One ---"
    let part01 = t' 40 s
    putStrLn $ "\t Your puzzle answer is " ++ show (length part01)

    putStrLn $ "--- Part Two ---"
    let part02 = t' 10 part01
    putStrLn $ "\t Your puzzle answer is " ++ show (length part02)

File: AoC/MMXV/rust/day10.rs
use std::{iter::Peekable, ops::Range, str::Chars};

use mmxv_rust::solved_output;

const PUZZLE: &str = include_str!("../puzzles/day10.txt");

fn repeat(input: &str, times: Range<u8>) -> String {
    times
        .into_iter()
        .fold(String::from(input), |las, _| look_and_say(&las))
}

fn look_and_say(input: &str) -> String {
    let mut output = String::new();
    let mut chars = input.chars().peekable();

    while let Some(ch) = chars.next() {
        let next_ch = lookup_until(&mut chars, |nch| ch.ne(nch)) + 1;
        output.push_str(&next_ch.to_string());
        output.push_str(&ch.to_string());
    }

    output
}

fn lookup_until(
    chars: &mut Peekable<Chars>,
    predicate: impl Fn(&char) -> bool,
) -> usize {
    let mut counter = 0;
    while let Some(ch) = chars.peek() {
        if predicate(ch) {
            break;
        }
        chars.next();
        counter += 1;
    }
    counter
}

fn main() {
    println!("--- Day 10: Elves Look, Elves Say ---");

    println!("--- Part One ---");
    let part01 = solve_part01(PUZZLE);
    println!("\tYour puzzle answer is {}.", solved_output(part01).len());
}

fn solve_part01(input: &str) -> String {
    repeat(input, 0..40)
}
